### PR TITLE
GSoC WIP: 2D hyperparameter visualization

### PR DESCRIPTION
**Pull request by MasonGallo**
- [x] base tests passed
- [x] WIP test added
- [x] WIP documentation added

As we discussed, this is part 1 with the initial WIP implementation only covering 2D plots aka 1 X and 1 Y only. This also contains the ability to "colorize" local optima. Very much a rough draft, so please lmk your feedback.

### CODE REVIEW

**Comment by larskotthoff** [comment]
I'm not sure about the name "validation" -- it's not really used to validate anything. Maybe "trajectory" instead? @berndbischl what do you think?

**Comment by zmjones** [comment]
what about HyperparameterPath?

**Comment by MasonGallo** [comment]
I don't love the name(s) either. FWIW, we called them validation curves where I worked. I will search through the literature to see if there's a more common name.

**Comment by MasonGallo** [comment]
Following up - I can't really seem to find a consistent "name" for these visualizations in the literature. Most don't really give it a name e.g. http://aad.informatik.uni-freiburg.de/papers/15-AAAI-MI-SMBO.pdf

Looking at the "competition", sklearn calls them validation curves while caret calls it resampling profile in their tutorial.

**Comment by berndbischl** [comment]
regarding the name: "ValidationCurve" I really dislike because it is so general. 
How about "plotHyperParamEffect"? Or "plotHyperParamInfluence"?
We might even make it more specific
plotHyperParamTuningEffect"? Or "plotHyperParamTuningInfluence

**Comment by MasonGallo** [comment]
Thanks all! Personally, I'm leaning toward plotHyperParamTuningEffect but we can discuss in our next meeting

**Comment by MasonGallo** [comment]
A quick big picture overview of the major use cases that this work will cover is available at: https://github.com/MasonGallo/model-optimization/tree/master/visualize_hyperparam_tuning/big_picture

Please let me know if you think of any major use cases that I've missed!

**Comment by zmjones** [comment]
you shouldn't need to reference checkmate functions with `::` since it is imported.

**Comment by berndbischl** [comment]
start rox title lines with "@title" please

**Comment by larskotthoff** [comment]
We converged on `plotHyperParsEffect`, for consistency and precision. Elegant, yet simple.

**Comment by MasonGallo** [comment]
Thanks all! Changes we discussed in today's meeting & from above:

- Focus on "1D" for this PR (although I suppose it's technically 2D)
- No reference to checkmate:: necessary
- Add @title for rox
- Argument name changes (x.axis, y.axis, extra.axis)
- Specify defaults
- Account for "tuning crashes" by replacing NAs in plotting method
- Account for lists of objects in generateData method in case of nested CV

**Comment by MasonGallo** [comment]
Not sure what's causing Travis to fail for `run-tune.R`. Any ideas?

**Comment by larskotthoff** [comment]
This was an unrelated failure caused by ParamHelpers and should be fixed now.

**Comment by MasonGallo** [comment]
Ping @berndbischl and @larskotthoff for feedback when you get a chance

**Comment by larskotthoff** [comment]
Thanks, looks good. Could you rename the R file to reflect the new name please?

**Comment by MasonGallo** [comment]
Thanks! The names are fixed.

**Comment by berndbischl** [comment]
please dont use TODO but FIXME

**Comment by berndbischl** [comment]
```
#' @param HyperParsEffectData [\code{HyperParsEffectData}]\cr
```

the arg name must be lower case and dot separated

**Comment by berndbischl** [comment]
```
		 +#' @return
 +#'  ggplot2 plot object
```

use template ret_ggplot2

**Comment by berndbischl** [comment]
try to always create a first block of arg check with checkmate in all functions.

**Comment by MasonGallo** [comment]
Changes we discussed in today's meeting:

- [ ] TODO -> FIXME
- [ ] fix argument name
- [ ] optimizer crash should read learner crash
- [ ] adjust learner crash depending on if measure is max/min
- [ ] nested cv should color lines/dots for each outer loop by default
- [ ] use cumulative for iteration plots so we don't show where global isn't improving
- [ ] rox fixes
- [ ] update markdown visualizations

https://github.com/MasonGallo/model-optimization/tree/master/visualize_hyperparam_tuning/big_picture

**Comment by MasonGallo** [comment]
The Travis failure looks to be a timeout on the regression tests, but that should not have any relation with this PR?

**Comment by larskotthoff** [comment]
Yeah, Travis sometimes doesn't like us. I've restarted that build.

**Comment by MasonGallo** [comment]
Reminder ping for @berndbischl

**Comment by berndbischl** [comment]
can you please rebase this on master so that we can see travis? 
i will try to review this now a bit more.

**Comment by berndbischl** [comment]
more review:

- please add an example for HyperParEffects with nested CV
- please dont have examples for the "generate" and the "plot" methods. have same in one place and refer the user to the other please by saying 
```
@examples
# see bla
```

**Comment by berndbischl** [comment]
final remarks.

- i guess this is fine. when @MasonGallo is done with above, please @larskotthoff also read once.

- @MasonGallo when merged, please do a PR for @schiffner in the tutorial to briefly explain users how this plots works

**Comment by MasonGallo** [comment]
@berndbischl 
Thanks! Above changes are done and I rebased / squashed into 1 commit. Looks like Jenkins is failing due to reason you mentioned in #973. I'll be in touch with @schiffner. 

I know NEWS suggestion is required - maybe we can discuss this in our meeting.

**Comment by larskotthoff** [comment]
Thanks, looks good. Merging.

**Comment by MasonGallo** [comment]
NEWS suggestion:

- added the ability to generate and plot data from the hyperparameter tuning path for a single hyperparameter

- new functions
-- generateHyperParsEffectData
-- plotHyperParsEffect

**Comment by larskotthoff** [comment]
Done, thanks.
